### PR TITLE
[ROCm][Bugfix] Fix Aiter RMSNorm 

### PR DESCRIPTION
**Pull request by vllmellm**
## Purpose

Currently in V1, RMSNorm was as CustomOps class.

The default behaviour of V1 when torch compile is enabled, the custom_ops is set to "none", this forces the rms norm class to call forward_native.

Since AITER RMS_NORM is faster, we will want to utilize the forward_cuda instead. So we have appended the "rms_norm" to the list `self.compilation_config.custom_ops` if user enable AITER RMS_NORM.

This PR fixes the AITER RMS NORM V1 custom op registration and added type checking as AITER RMSNORM does not support float32 inputs.

## Test Plan
Run Lmeval on a select models and compare aiter and non-aiter kernels.

## Test Result

Model | Task | Version | Filter | n-shot | Metric | Aiter On Value | Aiter Stderr | Aiter Off Value | Aiter Off Stderr
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
deepseek-ai/DeepSeek-V3 | gsm8k | 3 | flexible-extract | 5 | exact_match | 0.9530 | 0.0058 | 0.9492 | 0.0060
deepseek-ai/DeepSeek-V3 | gsm8k | 3 | strict-match | 5 | exact_match | 0.9515 | 0.0059 | 0.9492 | 0.0060
llama/Llama-4-Scout-17B-16E | gsm8k | 3 | flexible-extract | 5 | exact_match | 0.8211 | 0.0106 | 0.8271 | 0.0104
llama/Llama-4-Scout-17B-16E | gsm8k | 3 | strict-match | 5 | exact_match | 0.8203 | 0.0106 | 0.8256 | 0.0105
EmbeddedLLM/Qwen2.5-1.5B-Instruct-FP8-Dynamic | gsm8k | 3 | flexible-extract | 5 | exact_match | 0.5186 | 0.0138 | 0.5186 | 0.0138
EmbeddedLLM/Qwen2.5-1.5B-Instruct-FP8-Dynamic | gsm8k | 3 | strict-match | 5 | exact_match | 0.3397 | 0.0130 | 0.3404 | 0.0131 
Qwen/Qwen3-32B | gsm8k | 3 | flexible-extract | 5 | exact_match | 0.6224 | 0.0134 | 0.6240 | 0.0133
Qwen/Qwen3-32B | gsm8k | 3 | strict-match | 5 |exact_match | 0.7354 | 0.0122 | 0.7362 | 0.0121

## (Optional) Documentation Update
None
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/config/__init__.py [hunk 1]
```diff
@@ -3617,6 +3617,12 @@ def __post_init__(self):
         if self.compilation_config.pass_config.enable_sequence_parallelism:
             self.compilation_config.custom_ops.append("+rms_norm")
 
+        if envs.VLLM_USE_V1 and self.model_config is not None and \
+            not self.model_config.enforce_eager and current_platform.is_rocm():
+            if "none" in self.compilation_config.custom_ops:
+                self.compilation_config.custom_ops.remove("none")
```
Comment (ProExpertProg) [inline vllm/config/__init__.py]: We don't need to remove "none" from the list, "+rms_norm" overrides it. Also, could you move this to a platform-specific config update (`current_platform.check_and_update_config`)?

File: vllm/model_executor/layers/layernorm.py [hunk 2]
```diff
@@ -162,7 +196,7 @@ def forward_cuda(
             return self.forward_native(x, residual)
 
         add_residual = residual is not None
-        norm_func = dispatch_cuda_rmsnorm_func(add_residual)
+        norm_func = dispatch_rmsnorm_func(add_residual, self.dtype)
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/layernorm.py]: Instead of dispatching inside `forward_cuda`, please add `forward_rocm`. Also the dispatching can happen inside `__init__` instead of during runtime, that's the benefit of the `CustomOp` abstraction.

File: vllm/config/__init__.py (hunk 1 above)
Comment (ProExpertProg) [reply vllm/config/__init__.py]: Also, this enables the custom ROCm RmsNorm, are we sure we want to do that (even if AITER is not enabled)?

**Review by ProExpertProg — COMMENTED** [review]
Move some stuff around please otherwise looks good!

**Comment by ProExpertProg** [comment]
Also, @vllmellm could we do a follow-up where we enable aiter_rmsnorm + quant -> aiter_rmsnorm_quant fusion? I heard there were aiter kernels for this

File: vllm/config/__init__.py (hunk 1 above)
Comment (vllmellm) [reply vllm/config/__init__.py]: @ProExpertProg I moved it and added a check to append "+rms_norm" only when aiter is enabled

File: vllm/model_executor/layers/layernorm.py (hunk 2 above)
Comment (vllmellm) [reply vllm/model_executor/layers/layernorm.py]: @ProExpertProg i have moved the dispatch to the init. However, since ROCm and CUDA share the same logical flow, it would be redundant to implement the forward_hip with the same code. The CustomOp class assumes that HIP ops are compatible with CUDA ops. `forward_hip` defaults to the CUDA implementation and it is documented in the class. Perhaps we can implement a forward "helper" function and explicitly call it in both forward_cuda and forward_hip if that would make it clearer for other developers?

**Comment by vllmellm** [comment]
> Also, @vllmellm could we do a follow-up where we enable aiter_rmsnorm + quant -> aiter_rmsnorm_quant fusion? I heard there were aiter kernels for this

Sure thing! we'll just have to profile it first.

File: vllm/platforms/rocm.py [hunk 3]
```diff
@@ -323,23 +323,33 @@ def is_async_output_supported(cls, enforce_eager: Optional[bool]) -> bool:
...
                 parallel_config.worker_cls = "vllm.v1.worker.gpu_worker.Worker"
             else:
-                if envs.VLLM_USE_V1:
+                if use_v1:
                     parallel_config.worker_cls = \
                         "vllm.v1.worker.gpu_worker.Worker"
                 else:
                     parallel_config.worker_cls = "vllm.worker.worker.Worker"
 
+        if use_v1 and use_aiter_rms_norm and not enforce_eager:
```
Comment (ProExpertProg) [inline vllm/platforms/rocm.py]: Why does enforce_eager matter here?

File: vllm/model_executor/layers/layernorm.py [hunk 4]
```diff
@@ -162,13 +201,11 @@ def forward_cuda(
             return self.forward_native(x, residual)
 
         add_residual = residual is not None
-        norm_func = dispatch_cuda_rmsnorm_func(add_residual)
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/layernorm.py:165]: Can you still split up the CUDA and hip paths here? Even though there will be slight repetition, I think it will be more clear what's happening in each platform.

Comment (ProExpertProg) [reply vllm/model_executor/layers/layernorm.py:165]: (to be clear only hip should use the dispatching and CUDA should call the vllm custom ops directly)

File: vllm/model_executor/layers/layernorm.py [hunk 5]
```diff
@@ -114,6 +147,12 @@ def __init__(
             self.weight = torch.ones(hidden_size)
         if self.has_weight:
             self.weight = nn.Parameter(self.weight)
+        weight_dtype = self.weight.data.dtype
+
+        self.norm_func = dispatch_rmsnorm_func(with_fused_add=False,
+                                               dtype=weight_dtype)
+        self.norm_func_with_add = dispatch_rmsnorm_func(with_fused_add=True,
+                                                        dtype=weight_dtype)
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/layernorm.py]: Can you guard this under `is_rocm`?

**Review by ProExpertProg — CHANGES_REQUESTED** [review]

File: vllm/model_executor/layers/layernorm.py (hunk 2 above)
Comment (ProExpertProg) [reply vllm/model_executor/layers/layernorm.py]: > it would be redundant to implement the forward_hip with the same code

I think here separation of concerns is more important than DRY (don't repeat yourself).

> `forward_hip` defaults to the CUDA implementation and it is documented in the class

Yes, but here we're changing the default behavior, so it's okay to override the `forward_hip` method.

File: vllm/platforms/rocm.py (hunk 3 above)
Comment (vllmellm) [reply vllm/platforms/rocm.py]: This is solely for performance reasons. Aiter rms norm functions perform best when cuda graph capture is enabled.

Comment (ProExpertProg) [reply vllm/platforms/rocm.py]: In that case please add a comment and use the cudagraph_mode property: `cudagraph_mode != CUDAGraphMode.NONE`

File: vllm/model_executor/layers/layernorm.py [hunk 6]
```diff
@@ -114,6 +148,13 @@ def __init__(
             self.weight = torch.ones(hidden_size)
         if self.has_weight:
             self.weight = nn.Parameter(self.weight)
+        weight_dtype = self.weight.data.dtype
+
+        if current_platform.is_rocm():
+            self.rocm_aiter_norm_func = dispatch_rocm_rmsnorm_func(
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/layernorm.py]: Nit: `self.rocm_norm_fun`

**Review by ProExpertProg — COMMENTED** [review]
A quick nit and please avoid using enforce_eager directly

File: vllm/model_executor/layers/layernorm.py (hunk 6 above)
Comment (vllmellm) [reply vllm/model_executor/layers/layernorm.py]: @ProExpertProg done!

**Review by ProExpertProg — APPROVED** [review]
